### DISCUSSION TITLE
Import mapping in Remix build assets

### DISCUSSION THREAD

**Discussion by digitalsadhu** [remix-run/remix#3270] [Q&A]
### Some background:

For performance reasons, we use [import mapping](https://github.com/WICG/import-maps) to map bare imports to absolute URLs on our asset server. This gives us the ability to line up all our micro-frontend apps on the same version of React and gain perf benefits across the whole site (which is made up of many different apps). Users visit the site and download React and never download it again as they traverse the site (even though they are in practice visiting many pages and apps). Not doing this means that React is inlined in each Remix app and therefore the user is forced to download React again and again as they traverse the site. 

We publish our client side assets to an [Eik](https://eik.dev) server and typically do import mapping server side as part of our esbuild process. See [Eik import mapping](https://eik.dev/docs/mapping_import_map). We could however do this client side if that turns out to be an easier solution...

### What I've tried

I did some work patching Remix to support the Eik esbuild plugin which took the following import map:

https://assets.finn.no/map/react/2.0.1
```json
{
  "imports": {
    "react": "https://assets.finn.no/npm/react/v17/react.production.min.js",
    "react-dom": "https://assets.finn.no/npm/react-dom/v17/react-dom.production.min.js"
  }
}
```

and mapped statements like:
```js
import react from 'react';
```

to
```js
import react from 'https://assets.finn.no/npm/react/v17/react.production.min.js';
```

which is exactly what I want. Unfortunately, while the build succeeds, the app dies at runtime with the following error:

```
TypeError: Cannot read properties of undefined (reading 'element')
    at Routes (/src/objectpage-podium/src/remix/node_modules/@remix-run/react/components.js:169:75)
```

My guess is this likely has something to do with browser code being executed in a Node.js environment which does not support importing from absolute URLs like this: `import react from 'https://assets.finn.no/npm/react/v17/react.production.min.js';`. 

This is not normally an issue in our SSR React apps as we produce separate client and server bundles and only apply the mapping to code that will run in the browser.

### My questions:

1. Is it possible to externalise React from the build such that React is not inlined (and therefore we can apply import mapping in the browser while continuing to use react out of node_modules on the server)
2. Is is possible to force Remix to produce completely separate client and server bundles such that I can patch in import mapping to code that will only ever be run in the browser and not on the server while the server code continues to reference React out of node_modules?

Thanks for any thoughts/ideas!

**Comment by digitalsadhu** [comment]
I've now managed to get client side import mapping to work by patching the Remix Esbuild config

```
// @remix-run/dev/compiler.js line 331
-external: externals,
+external: ['react', 'react-dom', ...externals],
```

And then adding this import map to the document head

```jsx
<script type="importmap" dangerouslySetInnerHTML={ { __html: JSON.stringify({
    imports: {
        react: 'https://assets.finn.no/npm/react/v17/react.production.min.js',
        'react-dom': 'https://assets.finn.no/npm/react-dom/v17/react-dom.production.min.js',
        '@fabric-ds/react': 'https://assets.finn.no/pkg/@fabric-ds/react/v1/index.js'
    }
}) } }>
```

Is there was a way to tell Remix to externalise React without the need to patch the internals?

**Comment by pcattori** [comment]
Remix currently does not officially support import maps. You can read our [decision doc](https://github.com/remix-run/remix/blob/main/decisions/0001-use-npm-to-manage-npm-dependencies-for-deno-projects.md) for more context.

That said, even though we don't provide official support for import maps, we don't try to prevent you from doing so yourself. So it could be the case that this is possible through completely user-side code. But we haven't looked into it much.
